hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java

```diff
@@ -0,0 +1,1009 @@
+    this.containerBalancer = containerBalancer;
+    this.config = config;
+    this.metrics = metrics;
+    this.scmContext = scm.getScmContext();
+    this.overUtilizedNodes = new ArrayList<>();
+    this.underUtilizedNodes = new ArrayList<>();
+    this.withinThresholdUtilizedNodes = new ArrayList<>();
+    this.unBalancedNodes = new ArrayList<>();
+    this.placementPolicyValidateProxy = scm.getPlacementPolicyValidateProxy();
+    this.networkTopology = scm.getClusterMap();
+    this.nextIterationIndex = nextIterationIndex;
+    this.containerToSourceMap = new HashMap<>();
+    this.containerToTargetMap = new HashMap<>();
+    this.selectedSources = new HashSet<>();
+    this.selectedTargets = new HashSet<>();
+    findSourceStrategy = new FindSourceGreedy(nodeManager);
+  }
+
+  /**
+   * Balances the cluster in iterations. Regularly checks if balancing has
+   * been stopped.
+   */
+  public void run() {
+    try {
+      balancer();
+
+      // mark balancer completion, if ha and stopped as not a leader, it will
+      // be handled inside for not saving configuration with leader check
+      saveConfiguration(config, false, 0);
+    } catch (Throwable e) {
+      LOG.error("Container Balancer is stopped abnormally, ", e);
+    }
+    taskStatus = Status.STOPPED;
+  }
+
+  /**
+   * Tries to stop ContainerBalancer changing status to stopping. Calls
+   * {@link ContainerBalancerTask#stop()}.
+   */
+  public void stop() {
+    taskStatus = Status.STOPPING;
+  }
+
+  private void balancer() {
+    this.iterations = config.getIterations();
+    if (this.iterations == -1) {
+      //run balancer infinitely
+      this.iterations = Integer.MAX_VALUE;
+    }
+
+    // nextIterationIndex is the iteration that balancer should start from on
+    // leader change or restart
+    int i = nextIterationIndex;
+    for (; i < iterations && isBalancerRunning(); i++) {
+      // reset some variables and metrics for this iteration
+      resetState();
+
+      if (!isBalancerRunning()) {
+        return;
+      }
+
+      if (config.getTriggerDuEnable()) {
+        // before starting a new iteration, we trigger all the datanode
+        // to run `du`. this is an aggressive action, with which we can
+        // get more precise usage info of all datanodes before moving.
+        // this is helpful for container balancer to make more appropriate
+        // decisions. this will increase the disk io load of data nodes, so
+        // please enable it with caution.
+        nodeManager.refreshAllHealthyDnUsageInfo();
+        try {
+          long nodeReportInterval =
+              ozoneConfiguration.getTimeDuration(HDDS_NODE_REPORT_INTERVAL,
+                  HDDS_NODE_REPORT_INTERVAL_DEFAULT, TimeUnit.MILLISECONDS);
+          // one for sending command , one for running du, and one for
+          // reporting back make it like this for now, a more suitable
+          // value. can be set in the future if needed
+          wait(3 * nodeReportInterval);
```

Comment:
I think it is a requirement to have wait inside synchronized block.

Reply:
> I think it is a requirement to have wait inside synchronized block.

Not finding any usecase to be in synchronized block, since this is single run only. Searched all usecase in existing code.